# Bulk Geocoding Addresses

## Overview

We will be working with an Excel sheet containing addresses of Hurricane Evacuation Centers in New York City and geocode these addresses to create a map. We cover the following two methods for geocoding:

* *Geocoding Method 1*: Pandas and GeoPy: This is a more flexible method that uses [GeoPy](https://geopy.readthedocs.io/en/stable/) that offers more control on the process.
* *Geocoding Method 2*: GeoPandas: A simpler method [`geopandas.tools.geocode`](https://geopandas.org/en/stable/docs/reference/api/geopandas.tools.geocode.html) provided by GeoPandas that is easy to use, but with less options.

**Input Layers**:

* `Hurricane_Evacuation_Centers.xlsx`: A spreadsheet containing names and addresses of hurricane shelters inn NYC.

**Output**:

* `hurricane_evacuation_centers.zip`: A zipped shapefile with geocoded locations of the hurricane evacuation centers.

**Data Credit**: 

* [NYC Open Data Portal](https://data.cityofnewyork.us/Public-Safety/Hurricane-Evacuation-Centers-Map-/ayer-cga7)

**Watch Video Walkthrough** <a href='https://www.youtube.com/watch?v=N6itC6hbOvo' target='_blank'><img src='../images/yt_logo.png'></a>

## Setup and Data Download

The following blocks of code will install the required packages and download the datasets to your Colab environment.

In [73]:
import os
import re
import pandas as pd
import geopandas as gpd
import leafmap.foliumap as leafmap
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter
from zipfile import ZipFile

In [74]:
data_folder = 'data'
output_folder = 'output'

if not os.path.exists(data_folder):
    os.mkdir(data_folder)
if not os.path.exists(output_folder):
    os.mkdir(output_folder)

In [75]:
def download(url):
    filename = os.path.join(data_folder, os.path.basename(url))
    if not os.path.exists(filename):
        from urllib.request import urlretrieve
        local, _ = urlretrieve(url, filename)
        print('Downloaded ' + local)

data_url = 'https://github.com/spatialthoughts/geopython-tutorials/releases/download/data/'

download(data_url + 'Hurricane_Evacuation_Centers.xlsx')

## Procedure

Read the `Hurricane_Evacuation_Centers.xlsx` file using Pandas. This step requires the `openpyxl` package to be installed along with `pandas`. The dataset contains 64 addresses of hurricane evacuation centers in New York City.

In [96]:
excel_file = 'Hurricane_Evacuation_Centers.xlsx'
excel_file_path = os.path.join(data_folder, excel_file)
address_df = pd.read_excel(excel_file_path, skiprows=[0])
address_df

CITY                                     EC_Name  \
0           Flushing                         J.H.S. 185 - Queens   
1           Brooklyn                         I.S. 258 - Brooklyn   
2           Brooklyn                          I.S. 88 - Brooklyn   
3              Bronx                               Walton HS - X   
4           Brooklyn  P.S./I.S.30 Mary White Ovington - Brooklyn   
..               ...                                         ...   
59             Bronx                            I.S. 201 - Bronx   
60             Bronx                             P.S. 46 - Bronx   
61  Long Island City                             Aviation HS - Q   
62             Bronx                            P.S. 102 - Bronx   
63             Bronx  Morris Heights Educational Complex - Bronx   

                                 ADDRESS  ZIP_CODE  BOROCODE STATE ACCESSIBLE  
0                        147-26 25 Drive     11354         4    NY          Y  
1                       141 Macon Street     11216         3    NY          Y  
2                           544 7 Avenue     11215         3    NY          Y  
3                  2780 Reservoir Avenue     10468         2    NY          Y  
4                          7002 4 Avenue     11209         3    NY          Y  
..                                   ...       ...       ...   ...        ...  
59                     730 Bryant Avenue     10474         2    NY          Y  
60                    2760 Briggs Avenue     10458         2    NY          Y  
61                       45-30 36 Street     11101         4    NY          Y  
62                    1827 Archer Street     10460         2    NY          Y  
63  1780 Dr. Martin Luther King Jr. Blvd     10453         2    NY          Y  

[64 rows x 7 columns]

The Nominatim geocoder expects the addresses with certain format. We fix our source addresses by making the street numbers as ordinal numbers.

Examples

- `45-30 36 Street` → `45-30 36th Street`.
- `544 7 Avenue` → `544 7th Avenue`

In [97]:
def make_ordinal(match):
    n = int(match.group(1))
    if 11 <= (n % 100) <= 13:
        suffix = 'th'
    else:
        suffix = ['th', 'st', 'nd', 'rd', 'th'][min(n % 10, 4)]
    return str(n) + suffix + match.group(2)

def update_address(row):
  old_address = row['ADDRESS']
  pattern = r'(\d+)(\s+(?:Street|Avenue|Blvd|Drive))'
  result = re.sub(pattern, make_ordinal, old_address)
  return result

address_df['ADDRESS_FIXED'] = address_df.apply(update_address, axis=1)
address_df

CITY                                     EC_Name  \
0           Flushing                         J.H.S. 185 - Queens   
1           Brooklyn                         I.S. 258 - Brooklyn   
2           Brooklyn                          I.S. 88 - Brooklyn   
3              Bronx                               Walton HS - X   
4           Brooklyn  P.S./I.S.30 Mary White Ovington - Brooklyn   
..               ...                                         ...   
59             Bronx                            I.S. 201 - Bronx   
60             Bronx                             P.S. 46 - Bronx   
61  Long Island City                             Aviation HS - Q   
62             Bronx                            P.S. 102 - Bronx   
63             Bronx  Morris Heights Educational Complex - Bronx   

                                 ADDRESS  ZIP_CODE  BOROCODE STATE ACCESSIBLE  \
0                        147-26 25 Drive     11354         4    NY          Y   
1                       141 Macon Street     11216         3    NY          Y   
2                           544 7 Avenue     11215         3    NY          Y   
3                  2780 Reservoir Avenue     10468         2    NY          Y   
4                          7002 4 Avenue     11209         3    NY          Y   
..                                   ...       ...       ...   ...        ...   
59                     730 Bryant Avenue     10474         2    NY          Y   
60                    2760 Briggs Avenue     10458         2    NY          Y   
61                       45-30 36 Street     11101         4    NY          Y   
62                    1827 Archer Street     10460         2    NY          Y   
63  1780 Dr. Martin Luther King Jr. Blvd     10453         2    NY          Y   

                           ADDRESS_FIXED  
0                      147-26 25th Drive  
1                       141 Macon Street  
2                         544 7th Avenue  
3                  2780 Reservoir Avenue  
4                        7002 4th Avenue  
..                                   ...  
59                     730 Bryant Avenue  
60                    2760 Briggs Avenue  
61                     45-30 36th Street  
62                    1827 Archer Street  
63  1780 Dr. Martin Luther King Jr. Blvd  

[64 rows x 8 columns]

We also create a new column containing the full address. The `CITY` column in the data are actually boroughs of NYC and adding it causing many geocoding requests to fail. Instead, we use the city name 'NYC' for all addresses.

In [98]:
address_df['Full_Address'] = (
    address_df['ADDRESS_FIXED'] + ',' +
    'NYC' + ',' +
    address_df['STATE']+ ',' +
    address_df['ZIP_CODE'].astype(str))
address_df

CITY                                     EC_Name  \
0           Flushing                         J.H.S. 185 - Queens   
1           Brooklyn                         I.S. 258 - Brooklyn   
2           Brooklyn                          I.S. 88 - Brooklyn   
3              Bronx                               Walton HS - X   
4           Brooklyn  P.S./I.S.30 Mary White Ovington - Brooklyn   
..               ...                                         ...   
59             Bronx                            I.S. 201 - Bronx   
60             Bronx                             P.S. 46 - Bronx   
61  Long Island City                             Aviation HS - Q   
62             Bronx                            P.S. 102 - Bronx   
63             Bronx  Morris Heights Educational Complex - Bronx   

                                 ADDRESS  ZIP_CODE  BOROCODE STATE ACCESSIBLE  \
0                        147-26 25 Drive     11354         4    NY          Y   
1                       141 Macon Street     11216         3    NY          Y   
2                           544 7 Avenue     11215         3    NY          Y   
3                  2780 Reservoir Avenue     10468         2    NY          Y   
4                          7002 4 Avenue     11209         3    NY          Y   
..                                   ...       ...       ...   ...        ...   
59                     730 Bryant Avenue     10474         2    NY          Y   
60                    2760 Briggs Avenue     10458         2    NY          Y   
61                       45-30 36 Street     11101         4    NY          Y   
62                    1827 Archer Street     10460         2    NY          Y   
63  1780 Dr. Martin Luther King Jr. Blvd     10453         2    NY          Y   

                           ADDRESS_FIXED  \
0                      147-26 25th Drive   
1                       141 Macon Street   
2                         544 7th Avenue   
3                  2780 Reservoir Avenue   
4                        7002 4th Avenue   
..                                   ...   
59                     730 Bryant Avenue   
60                    2760 Briggs Avenue   
61                     45-30 36th Street   
62                    1827 Archer Street   
63  1780 Dr. Martin Luther King Jr. Blvd   

                                         Full_Address  
0                      147-26 25th Drive,NYC,NY,11354  
1                       141 Macon Street,NYC,NY,11216  
2                         544 7th Avenue,NYC,NY,11215  
3                  2780 Reservoir Avenue,NYC,NY,10468  
4                        7002 4th Avenue,NYC,NY,11209  
..                                                ...  
59                     730 Bryant Avenue,NYC,NY,10474  
60                    2760 Briggs Avenue,NYC,NY,10458  
61                     45-30 36th Street,NYC,NY,11101  
62                    1827 Archer Street,NYC,NY,10460  
63  1780 Dr. Martin Luther King Jr. Blvd,NYC,NY,10453  

[64 rows x 9 columns]

## Geocoding Method 1: Pandas

This is the preferred method that gives you more control over the geocoding process but involves a few more steps.

Now we use the [Nominatim Geocoder](https://geopy.readthedocs.io/en/stable/#nominatim) from GeoPy. Nominatim geocoder is based on OpenStreetMap data.

The geocoder is free to use but requires rate limiting our queries to 1 query per second. We use the `RateLimiter` provided by `geopy` package to set the delay. Since geocoding many address with the rate limit can take some time - we use a progress bar provided by the `tqdm` package to track the progress.

> You can also install and run Nominatim on your computer without any limitations and can geocode large number of addresses easily. Here's a good tutorial showing how to [Run Nominatim using Docker](https://nominatim.org/release-docs/latest/admin/Installation/).

In [116]:
from tqdm.notebook import tqdm

tqdm.pandas()

locator = Nominatim(user_agent='spatialthoughts', timeout=10)
geocode = RateLimiter(locator.geocode, min_delay_seconds=1)

address_df_pd = address_df.copy()
address_df_pd['location'] = address_df_pd['Full_Address'].progress_apply(geocode)
address_df_pd

  0%|          | 0/64 [00:00<?, ?it/s]

CITY                                     EC_Name  \
0           Flushing                         J.H.S. 185 - Queens   
1           Brooklyn                         I.S. 258 - Brooklyn   
2           Brooklyn                          I.S. 88 - Brooklyn   
3              Bronx                               Walton HS - X   
4           Brooklyn  P.S./I.S.30 Mary White Ovington - Brooklyn   
..               ...                                         ...   
59             Bronx                            I.S. 201 - Bronx   
60             Bronx                             P.S. 46 - Bronx   
61  Long Island City                             Aviation HS - Q   
62             Bronx                            P.S. 102 - Bronx   
63             Bronx  Morris Heights Educational Complex - Bronx   

                                 ADDRESS  ZIP_CODE  BOROCODE STATE ACCESSIBLE  \
0                        147-26 25 Drive     11354         4    NY          Y   
1                       141 Macon Street     11216         3    NY          Y   
2                           544 7 Avenue     11215         3    NY          Y   
3                  2780 Reservoir Avenue     10468         2    NY          Y   
4                          7002 4 Avenue     11209         3    NY          Y   
..                                   ...       ...       ...   ...        ...   
59                     730 Bryant Avenue     10474         2    NY          Y   
60                    2760 Briggs Avenue     10458         2    NY          Y   
61                       45-30 36 Street     11101         4    NY          Y   
62                    1827 Archer Street     10460         2    NY          Y   
63  1780 Dr. Martin Luther King Jr. Blvd     10453         2    NY          Y   

                           ADDRESS_FIXED  \
0                      147-26 25th Drive   
1                       141 Macon Street   
2                         544 7th Avenue   
3                  2780 Reservoir Avenue   
4                        7002 4th Avenue   
..                                   ...   
59                     730 Bryant Avenue   
60                    2760 Briggs Avenue   
61                     45-30 36th Street   
62                    1827 Archer Street   
63  1780 Dr. Martin Luther King Jr. Blvd   

                                         Full_Address  \
0                      147-26 25th Drive,NYC,NY,11354   
1                       141 Macon Street,NYC,NY,11216   
2                         544 7th Avenue,NYC,NY,11215   
3                  2780 Reservoir Avenue,NYC,NY,10468   
4                        7002 4th Avenue,NYC,NY,11209   
..                                                ...   
59                     730 Bryant Avenue,NYC,NY,10474   
60                    2760 Briggs Avenue,NYC,NY,10458   
61                     45-30 36th Street,NYC,NY,11101   
62                    1827 Archer Street,NYC,NY,10460   
63  1780 Dr. Martin Luther King Jr. Blvd,NYC,NY,10453   

                                             location  
0   (Junior High School 185, 147-26, 25th Drive, L...  
1   (Junior High School 258, 141, Macon Street, Be...  
2   (Middle School 88, 544, 7th Avenue, South Slop...  
3   (Celia Cruz High School of Music, 2780, Reserv...  
4   (7002, 4th Avenue, Bay Ridge, Brooklyn, Kings ...  
..                                                ...  
59  (730, Bryant Avenue, Hunts Point, Bronx County...  
60  (2760, Briggs Avenue, Bedford Park, Bronx Coun...  
61  (45-30, 36th Street, Sunnyside, Queens, City o...  
62  (Public School 102, 1827, Archer Street, Bronx...  
63  (East 125th Street, Manhattan Community Board ...  

[64 rows x 10 columns]

The response of the geocoding request is a [`geopy.location.Location`](https://geopy.readthedocs.io/en/stable/#geopy.location.Location) object containing the address along with geocoded coordinates as latitude and longitude. We extract them and add them to separate columns.

In [117]:
address_df_pd['latitude'] = address_df_pd['location'].apply(lambda loc: loc.latitude if loc else None)
address_df_pd['longitude'] = address_df_pd['location'].apply(lambda loc: loc.longitude if loc else None)
address_df_pd

CITY                                     EC_Name  \
0           Flushing                         J.H.S. 185 - Queens   
1           Brooklyn                         I.S. 258 - Brooklyn   
2           Brooklyn                          I.S. 88 - Brooklyn   
3              Bronx                               Walton HS - X   
4           Brooklyn  P.S./I.S.30 Mary White Ovington - Brooklyn   
..               ...                                         ...   
59             Bronx                            I.S. 201 - Bronx   
60             Bronx                             P.S. 46 - Bronx   
61  Long Island City                             Aviation HS - Q   
62             Bronx                            P.S. 102 - Bronx   
63             Bronx  Morris Heights Educational Complex - Bronx   

                                 ADDRESS  ZIP_CODE  BOROCODE STATE ACCESSIBLE  \
0                        147-26 25 Drive     11354         4    NY          Y   
1                       141 Macon Street     11216         3    NY          Y   
2                           544 7 Avenue     11215         3    NY          Y   
3                  2780 Reservoir Avenue     10468         2    NY          Y   
4                          7002 4 Avenue     11209         3    NY          Y   
..                                   ...       ...       ...   ...        ...   
59                     730 Bryant Avenue     10474         2    NY          Y   
60                    2760 Briggs Avenue     10458         2    NY          Y   
61                       45-30 36 Street     11101         4    NY          Y   
62                    1827 Archer Street     10460         2    NY          Y   
63  1780 Dr. Martin Luther King Jr. Blvd     10453         2    NY          Y   

                           ADDRESS_FIXED  \
0                      147-26 25th Drive   
1                       141 Macon Street   
2                         544 7th Avenue   
3                  2780 Reservoir Avenue   
4                        7002 4th Avenue   
..                                   ...   
59                     730 Bryant Avenue   
60                    2760 Briggs Avenue   
61                     45-30 36th Street   
62                    1827 Archer Street   
63  1780 Dr. Martin Luther King Jr. Blvd   

                                         Full_Address  \
0                      147-26 25th Drive,NYC,NY,11354   
1                       141 Macon Street,NYC,NY,11216   
2                         544 7th Avenue,NYC,NY,11215   
3                  2780 Reservoir Avenue,NYC,NY,10468   
4                        7002 4th Avenue,NYC,NY,11209   
..                                                ...   
59                     730 Bryant Avenue,NYC,NY,10474   
60                    2760 Briggs Avenue,NYC,NY,10458   
61                     45-30 36th Street,NYC,NY,11101   
62                    1827 Archer Street,NYC,NY,10460   
63  1780 Dr. Martin Luther King Jr. Blvd,NYC,NY,10453   

                                             location   latitude  longitude  
0   (Junior High School 185, 147-26, 25th Drive, L...  40.774912 -73.818619  
1   (Junior High School 258, 141, Macon Street, Be...  40.681918 -73.945597  
2   (Middle School 88, 544, 7th Avenue, South Slop...  40.660432 -73.987836  
3   (Celia Cruz High School of Music, 2780, Reserv...  40.870595 -73.897498  
4   (7002, 4th Avenue, Bay Ridge, Brooklyn, Kings ...  40.633625 -74.024038  
..                                                ...        ...        ...  
59  (730, Bryant Avenue, Hunts Point, Bronx County...  40.816040 -73.885264  
60  (2760, Briggs Avenue, Bedford Park, Bronx Coun...  40.867779 -73.890245  
61  (45-30, 36th Street, Sunnyside, Queens, City o...  40.743475 -73.929464  
62  (Public School 102, 1827, Archer Street, Bronx...  40.838106 -73.865790  
63  (East 125th Street, Manhattan Community Board ...  40.802551 -73.933098  

[64 rows x 12 columns]

Some addresses might have failed to geocode. Let's review them.

In [118]:
failed = address_df_pd[address_df_pd['location'].isna()]
failed

CITY              EC_Name           ADDRESS  ZIP_CODE  BOROCODE STATE  \
54  New York  P.S. 48 - Manhattan  4360-78 Broadway     10033         1    NY   

   ACCESSIBLE     ADDRESS_FIXED                   Full_Address location  \
54          Y  4360-78 Broadway  4360-78 Broadway,NYC,NY,10033     None   

    latitude  longitude  
54       NaN        NaN

We may discard these rows. But sometimes it is useful to manually geocode failed queries. You may use another service manually or obtain the coordinates from another source. The below block assigns the coordinates 40.85339,-73.93350 to the row with index 54.

In [119]:
address_df_pd.loc[54, ['latitude', 'longitude']] = (40.85339,-73.93350)

Select a subset of columns to keep in the final result and rename them.

In [120]:
address_df_pd = address_df_pd[['EC_Name', 'Full_Address', 'latitude', 'longitude']]
address_df_pd.rename(columns = {'EC_Name': 'Name', 'Full_Address': 'Address'}, inplace=True)

We create a GeoDataFrame from the latitude and longitude columns.

In [121]:
geometry = gpd.points_from_xy(address_df_pd.longitude, address_df_pd.latitude)
address_gdf = gpd.GeoDataFrame(address_df_pd, crs='EPSG:4326', geometry=geometry)
address_gdf

Name  \
0                          J.H.S. 185 - Queens   
1                          I.S. 258 - Brooklyn   
2                           I.S. 88 - Brooklyn   
3                                Walton HS - X   
4   P.S./I.S.30 Mary White Ovington - Brooklyn   
..                                         ...   
59                            I.S. 201 - Bronx   
60                             P.S. 46 - Bronx   
61                             Aviation HS - Q   
62                            P.S. 102 - Bronx   
63  Morris Heights Educational Complex - Bronx   

                                              Address   latitude  longitude  \
0                      147-26 25th Drive,NYC,NY,11354  40.774912 -73.818619   
1                       141 Macon Street,NYC,NY,11216  40.681918 -73.945597   
2                         544 7th Avenue,NYC,NY,11215  40.660432 -73.987836   
3                  2780 Reservoir Avenue,NYC,NY,10468  40.870595 -73.897498   
4                        7002 4th Avenue,NYC,NY,11209  40.633625 -74.024038   
..                                                ...        ...        ...   
59                     730 Bryant Avenue,NYC,NY,10474  40.816040 -73.885264   
60                    2760 Briggs Avenue,NYC,NY,10458  40.867779 -73.890245   
61                     45-30 36th Street,NYC,NY,11101  40.743475 -73.929464   
62                    1827 Archer Street,NYC,NY,10460  40.838106 -73.865790   
63  1780 Dr. Martin Luther King Jr. Blvd,NYC,NY,10453  40.802551 -73.933098   

                      geometry  
0   POINT (-73.81862 40.77491)  
1   POINT (-73.94560 40.68192)  
2   POINT (-73.98784 40.66043)  
3   POINT (-73.89750 40.87059)  
4   POINT (-74.02404 40.63362)  
..                         ...  
59  POINT (-73.88526 40.81604)  
60  POINT (-73.89025 40.86778)  
61  POINT (-73.92946 40.74347)  
62  POINT (-73.86579 40.83811)  
63  POINT (-73.93310 40.80255)  

[64 rows x 5 columns]

Visualize the geocoded addresses on a map.

In [122]:
import folium
m = leafmap.Map(width=800, height=500)
address_gdf.explore(
    m=m,
    marker_type='marker',
    marker_kwds={
        'icon': folium.Icon(color='#fdbb84', icon='hurricane', prefix='fa')
    }
)
m.zoom_to_gdf(address_gdf)
m

Save the results as a shapefile.

In [123]:
output_file = 'hurricane_evacuation_centers.shp'
output_path = os.path.join(output_folder, output_file)

address_gdf.to_file(filename=output_path)

We also zip the resulting shapefile along with all sidecar files so it can be downloaded easily.

In [124]:
output_zip_file = 'hurricane_evacuation_centers.zip'
output_zip_path = os.path.join(output_folder, output_zip_file)

sidecar_files = [
    os.path.join(output_folder, file)
    for file in os.listdir(output_folder)
    if file.endswith(('shp', 'shx', 'dbf', 'prj'))
]

with ZipFile(output_zip_path, 'w') as zip_object:
    for sidecar in sidecar_files:
      zip_object.write(sidecar, os.path.basename(sidecar))


## Geocoding Method 2: GeoPandas


GeoPandas also provides a helper tool [geopandas.tools.geocode](https://geopandas.org/en/stable/docs/reference/api/geopandas.tools.geocode.html) for geocoding that allows using a GeoPy geocoder. This is a handy tool that gives you a GeoDataFrame but doesn't offer the flexibility of the previous method. We demonstrate the process here for the sake of completeness.

In [106]:
locator = Nominatim(user_agent='spatialthoughts', timeout=10)
geocode = RateLimiter(locator.geocode, min_delay_seconds=1)

address_df_gpd = address_df.copy()

gdf = gpd.tools.geocode(address_df_gpd['ADDRESS_FIXED'])
gdf

geometry  \
0    POINT (-73.81864 40.77489)   
1    POINT (-88.95624 39.84059)   
2   POINT (-122.19749 47.68078)   
3    POINT (-73.89657 40.87026)   
4    POINT (-73.98978 40.67027)   
..                          ...   
59   POINT (-73.88526 40.81604)   
60   POINT (-93.49384 34.93464)   
61   POINT (-73.92946 40.74347)   
62   POINT (-73.86579 40.83811)   
63   POINT (-80.24120 25.83239)   

                                              address  
0   147-26, 25th Drive, 11354, 25th Drive, New Yor...  
1   Macon County Office Building, 141, South Main ...  
2   544, 7th Avenue, 98033, 7th Avenue, Kirkland, ...  
3   2780, Reservoir Avenue, 10468, Reservoir Avenu...  
4   4th Avenue, 10th Street, 11215, 10th Street, N...  
..                                                ...  
59  730, Bryant Avenue, 10474, Bryant Avenue, New ...  
60  Briggsville Avenue, 72828, Arkansas, United St...  
61  45-30, 36th Street, 11101, 36th Street, New Yo...  
62  Public School 102, 1827, Archer Street, 10460,...  
63  Dr. Martin Luther King, Jr., Northwest 27th Av...  

[64 rows x 2 columns]

We merge the resulting GeoDataFrame with the original DataFrame to retain some columns from the original dataset.

In [ ]:
merged = gdf.join(address_df_gpd)
merged

We select the subset of columns and rename them.

In [109]:
merged = merged[['EC_Name', 'Full_Address', 'geometry']]
merged.rename(columns = {'EC_Name': 'Name', 'Full_Address': 'Address'}, inplace=True)
merged

Name  \
0                          J.H.S. 185 - Queens   
1                          I.S. 258 - Brooklyn   
2                           I.S. 88 - Brooklyn   
3                                Walton HS - X   
4   P.S./I.S.30 Mary White Ovington - Brooklyn   
..                                         ...   
59                            I.S. 201 - Bronx   
60                             P.S. 46 - Bronx   
61                             Aviation HS - Q   
62                            P.S. 102 - Bronx   
63  Morris Heights Educational Complex - Bronx   

                                              Address  \
0                      147-26 25th Drive,NYC,NY,11354   
1                       141 Macon Street,NYC,NY,11216   
2                         544 7th Avenue,NYC,NY,11215   
3                  2780 Reservoir Avenue,NYC,NY,10468   
4                        7002 4th Avenue,NYC,NY,11209   
..                                                ...   
59                     730 Bryant Avenue,NYC,NY,10474   
60                    2760 Briggs Avenue,NYC,NY,10458   
61                     45-30 36th Street,NYC,NY,11101   
62                    1827 Archer Street,NYC,NY,10460   
63  1780 Dr. Martin Luther King Jr. Blvd,NYC,NY,10453   

                       geometry  
0    POINT (-73.81864 40.77489)  
1    POINT (-88.95624 39.84059)  
2   POINT (-122.19749 47.68078)  
3    POINT (-73.89657 40.87026)  
4    POINT (-73.98978 40.67027)  
..                          ...  
59   POINT (-73.88526 40.81604)  
60   POINT (-93.49384 34.93464)  
61   POINT (-73.92946 40.74347)  
62   POINT (-73.86579 40.83811)  
63   POINT (-80.24120 25.83239)  

[64 rows x 3 columns]